File: lintcode/swap_nodes_in_pairs/cpp/swap_nodes_in_pairs.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/swap-nodes-in-pairs
@Language: C++
@Datetime: 17-02-16 00:56
*/

#include <stack>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head a ListNode
     * @return a ListNode
     */
    ListNode* swapPairs(ListNode* head) 
    {
        auto stack = std::stack<ListNode*>{ };
        auto dummy = ListNode{ 0 };
        auto tail  = &dummy;
        auto curr  = head;
        
        while (curr != nullptr)
        {
            stack.push(curr);
            
            curr = curr->next;
            
            if (curr == nullptr || stack.size() == 2) 
            {
                while (stack.size() > 0) {
                    tail->next = stack.top();
                    tail       = tail->next;
                    stack.pop();
                }
                
                tail->next = nullptr;
            }
        }
        
        return dummy.next;
    }
};

File: lintcode/easy/swap_nodes_in_pairs/py/swap_nodes_in_pairs.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/swap-nodes-in-pairs
@Language: Python
@Datetime: 17-02-16 00:32
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head == None:
            return None
            
        dummy  = ListNode(0)
        dummy.next = head
    
        prev   = dummy
        first  = dummy
        second = None
            
        while first.next != None and first.next.next != None:
            first       = first.next
            second      = first.next
    
            temp        = second.next
                
            second.next = first
            first.next  = temp
                
            prev.next   = second
            prev        = first
                
        return dummy.next